Sources/NIOCore/Docs.docc/ByteBuffer-lengthPrefix.md

```diff
@@ -0,0 +1,119 @@
+///     - strategy: The strategy to use for encoding the length.
+///     - writeData: A closure that takes a buffer, writes some data to it, and returns the number of bytes written.
+/// - Returns: Number of total bytes written. This is the length of the written data + the number of bytes used to write the length before it.
+public mutating func writeLengthPrefixed<Strategy: NIOBinaryIntegerEncodingStrategy>(
+    strategy: Strategy,
+    writeData: (_ buffer: inout ByteBuffer) throws -> Int
+) rethrows -> Int
+```
+
+Users could use the function as follows:
+
+```swift
+myBuffer.writeLengthPrefixed(strategy: .quic) { writer in
+    writer.writeString("something")
+    writer.writeSomethingComplex(something)
+}
+```
+
+Writing the implementation of `writeLengthPrefixed` presented a challenge. We need to write the length _before_ the
+data. But we do not know the length until the data is written.
+
+Ideally, we would reserve some number of bytes, then call the writeData closure, and then go back and write the length
+in the reserved space. However, we would not even know how many bytes of space to reserve, because the number of bytes
+needed to write an integer will depend on the integer!
+
+The solution we landed on was the following:
+
+- Added ``NIOBinaryIntegerEncodingStrategy/requiredBytesHint``. This allows strategies to provide an estimate of how
+  many bytes they need for encoding a length
+- Using this property, reserve the estimated number of bytes
+- Call the `writeData` closure to write the data
+- Go back to the reserved space to write the length
+    - If the length ends up needing fewer bytes than we had reserved, shuffle the data back to close the gap
+    - If the length ends up needing more bytes than we had reserved, shuffle the data forward to make space
+
+This code will be most performant when the `requiredBytesHint` is exactly correct, because it will avoid needing to
+shuffle any bytes. With that in mind, we can actually make one more optimisation: when we call the writeInteger function
+on a strategy, we can tell the strategy that we have already reserved some number of bytes. Some encoding strategies
+will be able to adjust the way they encode such that they can use exactly that many bytes.
+
+We added the following function to the ``NIOBinaryIntegerEncodingStrategy`` protocol. This is optional to implement, and
+will default to simply calling the existing ``NIOBinaryIntegerEncodingStrategy/writeInteger(_:to:)`` function.
+
+```swift
+/// - Parameters:
+///   - integer: The integer to write
+///   - reservedCapacity: The capacity already reserved for writing this integer
+///   - buffer: The buffer to write into.
+/// - Returns: The number of bytes used to write the integer.
+func writeInteger(
+    _ integer: Int,
+    reservedCapacity: Int,
+    to buffer: inout ByteBuffer
+) -> Int
+```
+
+Many strategies will not be able to do anything useful with the additional `reservedCapacity` parameter. For e.g., in
```

Comment:
nit: "e.g." is the abbreviation of "exempli gratia" which means "for the sake of example", so no need to write "For e.g.", Just use "e.g." or "for example"

```suggestion
Many strategies will not be able to do anything useful with the additional `reservedCapacity` parameter. For example, in
```